pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -980,14 +1073,45 @@ func (a *Actuator) remediateIfNeeded(ctx context.Context, machine *machinev1beta
 			return a.deleteNode(ctx, node)
 		}
 
-		// node is deleted, we can power on the host
+		//we need this annotation to differentiate between unhealthy machine that
+		//needs remediation, and an unhealthy machine that just got remediated
+		return a.addPoweredOffForRemediationAnnotation(ctx, machine)
+	}
+
+	// here we know that host has been powered off and node has been deleted
+	if hasPowerOffRequestAnnotation(baremetalhost) {
+		// we can now power host back on
 		log.Printf("Requesting Host %s power on for Machine %s",
 			baremetalhost.Name, machine.Name)
-		return a.requestPowerOn(ctx, baremetalhost)
+		return a.requestPowerOn(ctx, machine, baremetalhost)
 	}
 
+	node, err := a.getNodeByMachine(ctx, machine)
+
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			log.Printf("Failed to get Node from Machine %s: %s", machine.Name, err.Error())
+			return err
+		}
+	}
 	//node is still not running, so we requeue
 	if node == nil {
+		if canReprovision(machine, baremetalhost) {
+			mhc := a.getMhcByMachine(machine)
+			timeout := remediationPowerOnTimeout
+			if mhc != nil && mhc.Spec.NodeStartupTimeout.Duration != 0*time.Second {
+				timeout = mhc.Spec.NodeStartupTimeout.Duration
+			}
+			if isActionTimedOut(
+				machine,
+				powerOnRequestTimestampAnnotation,
+				timeout,
+			) {
+				log.Printf("Remediation (power on action) of machine %q takes longer than configured timeout %q. Deleting the machine.", machine.Name, timeout)
+				return a.tryDeleteMachine(ctx, machine)
+
+			}
+		}
```

Comment:
This function is already very long, could we split this whole block into a separate function maybe?